### PR TITLE
Validate pet set scale up/down order

### PR DESCRIPTION
**Pull request by dshulyak**
This change implements test cases to validate that:
- scale up will be done in order
- scale down in reverse order
- if any of the pets will be unhealthy, scale up/down will halt

related: https://github.com/kubernetes/kubernetes/issues/30082

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32868)

### CODE REVIEW

**Comment by bprashanth** [comment]
Great thanks, I'll try to take a look once we have 1.4 out (O(soon))

**Comment by dshulyak** [comment]
@bprashanth Thanks, i think it is ready for review. I will implement the last test case as a separate patch.

**Comment by dshulyak** [comment]
@bprashanth Please review when you will have time

File: test/e2e/petset.go
Comment (bprashanth) [inline test/e2e/petset.go]: aren't there standard api events you can use for these instead of introducing a dependency?

File: test/e2e/petset.go
Comment (bprashanth) [inline test/e2e/petset.go]: this is too easy and we know it works from all the other tests. Basically any test that invokes `saturate` does exactly this. 

Can you actually force a pet to become unhealthy? i.e it needs to fail a readiness probe or crashloop. The easiest way to do this is probably run nginx sidecar (or just as the pet) serving a file (as the readiness probe endpoint). remove the file with `kubectl exec` and it serves some non-200 code, put the file back and it's happy.

File: test/e2e/petset.go
Comment (bprashanth) [inline test/e2e/petset.go]: can you please add one that does a scale up, scale down, scale up (without waiting for anything inbetween), and checks that we converge to the final replica count in order? (i.e going from 3->5, 5->2, 2->4 should not ever create pets 4,3 in that order)

File: test/e2e/petset.go
Comment (bprashanth) [inline test/e2e/petset.go]: let me know if this suggestion isn't clear and I can elaborate

File: test/e2e/petset.go
Comment (dshulyak) [inline test/e2e/petset.go]: i will check, i just noticed it was the approach in another test

File: test/e2e/petset.go
Comment (dshulyak) [inline test/e2e/petset.go]: done

Comment (bprashanth) [inline test/e2e/petset.go]: I meant don't use the annotation at all to trigger "unhealthy-pet", instead: 
* bring up a petset
* trigger pet-X to fail a readiness probe (by doing the nginx idea, nginx serves file content of /tmp/foo, kubectl exec rm /tmp/foo will cause it to return 500, kubelet will probe for readiness and notice this marking the pet !ready)
* Wait till this surfaces as unready pod
* Try scaling up the petset
* Verify petset is blocked on actually unready pet 
* trigger pet-X to pass readiness probe (touch /tmp/foo)
... and so on

Comment (dshulyak) [inline test/e2e/petset.go]: by done i meant that it was clear from the first comment, i changed it in my branch and will push as soon as it will be ready

File: test/e2e/petset.go
Comment (dshulyak) [inline test/e2e/petset.go]: i wasn't able to find relevant events declared in api, do you have something specific in mind?

**Comment by dshulyak** [comment]
@bprashanth refactored a change according to discussion. I used /index.html for a readinessProbe, this way it requires less changes. Please take a look when you will have time

File: test/e2e/petset.go
Comment (janetkuo) [inline test/e2e/petset.go]: nit: put more info in test descriptor, see https://github.com/kubernetes/kubernetes/blob/master/docs/devel/writing-good-e2e-tests.md#some-concrete-tools-at-your-disposal

**Comment by dshulyak** [comment]
@janetkuo thanks for review, updated test descriptor

**Comment by dshulyak** [comment]
@k8s-bot gci gce e2e test this

**Comment by dshulyak** [comment]
@bprashanth @janetkuo maybe it can be merged?

**Comment by janetkuo** [comment]
@bprashanth are you still reviewing this one?

**Comment by bprashanth** [comment]
@janetkuo if you're ready to lgtm please go ahead

File: test/e2e/petset.go
Comment (bprashanth) [inline test/e2e/petset.go]: nit: which function is this (I think there's a similar function in framework.ExpectNoError)? is there a reason it needs to be global to the e2e package if it's a petset specific thing ? (sorry this is probably my fault but I just noticed it in this pr)

**Review by bprashanth — COMMENTED** [review]
Sorry for the delay, just had one question but LGTM (even if you don't fix the question, or want to address it but in a follow up pr or something). You do however have a rebase conflict, I think.

File: test/e2e/petset.go
Comment (janetkuo) [inline test/e2e/petset.go]: Do we still need this?

**Comment by janetkuo** [comment]
Things to consider for rebase:

1. we rename petset to statefulset and moved it to beta
2. initialized annotation behavior is flipped

**Comment by janetkuo** [comment]
Ping @dshulyak for rebase. Would love to have this test for 1.5

**Comment by dshulyak** [comment]
@janetkuo thanks for ping) will rebase it tomorrow in my tz

**Comment by dshulyak** [comment]
@janetkuo please take a look, rebased and used simple watch instead of controller

File: test/e2e/petset.go
Comment (dshulyak) [inline test/e2e/petset.go]: it doesn't test anything new actually, will remove

File: test/e2e/petset.go
Comment (janetkuo) [inline test/e2e/petset.go]: Scaling

Comment (janetkuo) [inline test/e2e/petset.go]: will run **in namespace**

File: test/e2e/petset.go
Comment (janetkuo) [inline test/e2e/petset.go]: Capitalize the first word of all "By" steps

**Comment by janetkuo** [comment]
Only nits, LGTM otherwise. 

We should probably remove `framework.SkipUnlessProviderIs("gce", "vagrant")` in "Pet set recreate" test as well (this can be a follow-up PR)

File: test/e2e/petset.go
Comment (janetkuo) [inline test/e2e/petset.go]: One more question: is 10s long enough?

**Comment by dshulyak** [comment]
@janetkuo fixed comments

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
Thanks!

**Pull request merged**
